arelle/plugin/validate/NL/rules/fr_nl.py

```diff
@@ -297,6 +297,119 @@ def rule_fr_nl_2_07(
+    """
+    for context in val.modelXbrl.contexts.values():
+        elt_start = XmlUtil.child(context.period, XbrlConst.xbrli, "startDate")
+        elt_end = XmlUtil.child(context.period, XbrlConst.xbrli, "endDate")
+        if ((elt_start is not None and 'T' in getattr(elt_start, 'text')) or
+                (elt_end is not None and 'T' in getattr(elt_end, 'text'))):
+            yield Validation.error(
+                codes='NL.FR-NL-3.01',
+                msg=_('Date elements in an \'xbrli:period\' element must be included without time'),
+                modelObject=context
+            )
+
+
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+    disclosureSystems=[
+        DISCLOSURE_SYSTEM_NT16,
+        DISCLOSURE_SYSTEM_NT17,
+        DISCLOSURE_SYSTEM_NT18
+    ],
+)
+def rule_fr_nl_3_02(
+        pluginData: PluginValidationDataExtension,
+        val: ValidateXbrl,
+        *args: Any,
+        **kwargs: Any,
+) -> Iterable[Validation] | None:
+    """
+    FR-NL-3.02: The element 'xbrli:forever' MUST NOT be used
+    """
+    for context in val.modelXbrl.contexts.values():
+        if context.isForeverPeriod:
+            yield Validation.error(
+                codes='NL.FR-NL-3.02',
+                msg=_('The element \'xbrli:forever\' must not be used'),
+                modelObject=context
+            )
+
+
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+    disclosureSystems=[
+        DISCLOSURE_SYSTEM_NT16,
+        DISCLOSURE_SYSTEM_NT17,
+        DISCLOSURE_SYSTEM_NT18
+    ],
+)
+def rule_fr_nl_3_03(
+        pluginData: PluginValidationDataExtension,
+        val: ValidateXbrl,
+        *args: Any,
+        **kwargs: Any,
+) -> Iterable[Validation] | None:
+    """
+    FR-NL-3.03: An XBRL instance document MUST NOT contain unused contexts
+    """
+    unused_contexts = list(set(val.modelXbrl.contexts.values()) - set(val.modelXbrl.contextsInUse))
+    unused_contexts.sort(key=lambda x: x.id)
+    for context in unused_contexts:
+        yield Validation.error(
+            codes='NL.FR-NL-3.03',
+            msg=_('Unused context must not exist in XBRL instance document'),
+            modelObject=context
+        )
+
+
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+    disclosureSystems=[
+        DISCLOSURE_SYSTEM_NT16,
+        DISCLOSURE_SYSTEM_NT17,
+        DISCLOSURE_SYSTEM_NT18
+    ],
+)
+def rule_fr_nl_4_02(
+        pluginData: PluginValidationDataExtension,
+        val: ValidateXbrl,
+        *args: Any,
+        **kwargs: Any,
+) -> Iterable[Validation] | None:
+    """
+    FR-NL-4.02: An XBRL instance document MUST NOT contain unused 'xbrli:unit' elements
+    """
+    unused_units_set = set(val.modelXbrl.units.values()) - {fact.unit for fact in val.modelXbrl.facts if fact.unit is not None}
+    unused_units = list(unused_units_set)
+    unused_units.sort(key=lambda x: x.hash)
+    for unit in unused_units:
+        yield Validation.error(
```

Comment:
Could this also be done with a single error similar to rule_fr_nl_2_04?
https://github.com/Arelle/Arelle/blob/d1662db3eae1aea6cdcfbf8fc4a4df5d74eb2dcf/arelle/plugin/validate/NL/rules/fr_nl.py#L189-L194

Reply:
Talked offline and decided this is good to go.